Detectors/CPV/calib/src/GainCalibrator.cxx

```diff
@@ -0,0 +1,281 @@
+void AmplitudeSpectrum::fillBinData(TH1F* h)
+{
+  for (uint16_t i = 0; i < nBins; i++) {
+    h->SetBinContent(i + 1, float(mBinContent[i]));
+    h->SetBinError(i + 1, sqrt(float(mBinContent[i])));
+  }
+}
+//_____________________________________________________________________________
+// GainCalibData
+//_____________________________________________________________________________
+void GainCalibData::fill(const gsl::span<const Digit> digits)
+{
+  for (auto& dig : digits) {
+    mAmplitudeSpectra[dig.getAbsId()].fill(dig.getAmplitude());
+  }
+}
+//_____________________________________________________________________________
+void GainCalibData::merge(const GainCalibData* prev)
+{
+  for (int i = 0; i < Geometry::kNCHANNELS; i++) {
+    mAmplitudeSpectra[i] += prev->mAmplitudeSpectra[i];
+  }
+  LOG(info) << "Merged GainCalibData with previous one.";
+  print();
+}
+//_____________________________________________________________________________
+void GainCalibData::print() const
+{
+  int nNonEmptySpectra = 0;
+  uint64_t nTotalEntries = 0;
+  for (int i = 0; i < Geometry::kNCHANNELS; i++) {
+    if (mAmplitudeSpectra[i].getNEntries()) {
+      nNonEmptySpectra++;
+      nTotalEntries += mAmplitudeSpectra[i].getNEntries();
+    }
+  }
+  LOG(info) << "GainCalibData::print() : "
+            << "I have " << nNonEmptySpectra
+            << " non-empty amplitude spectra with " << nTotalEntries
+            << " entries total ranged in (" << AmplitudeSpectrum::lRange << "; " << AmplitudeSpectrum::rRange << ").";
+}
+//_____________________________________________________________________________
+// GainCalibrator
+//_____________________________________________________________________________
+GainCalibrator::GainCalibrator()
+{
+  LOG(info) << "GainCalibrator::GainCalibrator() : "
+            << "Gain calibrator created!";
+  auto& cpvParams = o2::cpv::CPVCalibParams::Instance();
+  mMinEvents = cpvParams.mGainMinEvents;
+  mMinNChannelsToCalibrate = cpvParams.mGainMinNChannelsToCalibrate;
+  mDesiredLandauMPV = cpvParams.mGainDesiredLandauMPV;
+  mToleratedChi2PerNDF = cpvParams.mGainToleratedChi2PerNDF;
+  mMinAllowedCoeff = cpvParams.mGainMinAllowedCoeff;
+  mMaxAllowedCoeff = cpvParams.mGainMaxAllowedCoeff;
+  mFitRangeL = cpvParams.mGainFitRangeL;
+  mFitRangeR = cpvParams.mGainFitRangeR;
+  // adjust fit ranges to descrete binned values
+  if (mFitRangeL < AmplitudeSpectrum::lRange) {
+    mFitRangeL = AmplitudeSpectrum::lRange;
+  }
+  if (mFitRangeR > AmplitudeSpectrum::rRange) {
+    mFitRangeR = AmplitudeSpectrum::rRange;
+  }
+  double binWidth = (AmplitudeSpectrum::rRange - AmplitudeSpectrum::lRange) / AmplitudeSpectrum::nBins;
+  mFitRangeL = AmplitudeSpectrum::lRange + std::floor((mFitRangeL - AmplitudeSpectrum::lRange) / binWidth) * binWidth;
+  mFitRangeR = AmplitudeSpectrum::lRange + std::floor((mFitRangeR - AmplitudeSpectrum::lRange) / binWidth) * binWidth;
+  LOG(info) << "Parameters used: ";
```

Comment:
note that the cpvParams (as ConfigurableParam) has printKeyValues method, which can print also the provenance.